src/borg/testsuite/helpers_test.py

```diff
@@ -1408,6 +1408,85 @@ def test_passphrase_new_retries(self, monkeypatch):
     def test_passphrase_repr(self):
         assert "secret" not in repr(Passphrase("secret"))
 
+    def test_passphrase_wrong_debug_enabled(self, capsys, monkeypatch):
+        monkeypatch.setenv("BORG_DEBUG_PASSPHRASE", "YES")
+        monkeypatch.setenv("BORG_PASSPHRASE", "wrong_passphrase")
+        monkeypatch.setenv("BORG_PASSCOMMAND", "echo wrong_passphrase")
+        monkeypatch.setenv("BORG_PASSPHRASE_FD", "123")
+
+        with pytest.raises(PassphraseWrong) as exc_info:
+            Passphrase.display_debug_info("wrong_passphrase")
+            raise PassphraseWrong()
+
+        out, err = capsys.readouterr()
+        assert "Incorrect passphrase!" in err
+        assert 'Passphrase used (between double-quotes): "wrong_passphrase"' in err
+        assert "77726f6e675f70617373706872617365" in err
+        assert "BORG_PASSPHRASE" in err
+        assert "BORG_PASSCOMMAND" in err
+        assert "BORG_PASSPHRASE_FD" in err
+
+        assert str(exc_info.value) == (
+            "passphrase supplied in BORG_PASSPHRASE, by BORG_PASSCOMMAND or via BORG_PASSPHRASE_FD is incorrect."
+        )
+
+    def test_passphrase_wrong_debug_disabled(self, capsys, monkeypatch):
+        monkeypatch.delenv("BORG_DEBUG_PASSPHRASE", raising=False)
+
+        with pytest.raises(PassphraseWrong):
+            Passphrase.display_debug_info("wrong_passphrase")
+            raise PassphraseWrong()
+
+        out, err = capsys.readouterr()
+        assert "Incorrect passphrase!" not in err
+        assert 'Passphrase used (between double-quotes): "wrong_passphrase"' not in err
+
+    def test_verification(self, capsys, monkeypatch):
+        passphrase = "test_passphrase"
+
+        monkeypatch.setenv("BORG_DISPLAY_PASSPHRASE", "no")
+        Passphrase.verification(passphrase)
+        out, err = capsys.readouterr()
+        assert "Your passphrase (between double-quotes)" not in err
+
+        monkeypatch.setenv("BORG_DISPLAY_PASSPHRASE", "yes")
+        Passphrase.verification(passphrase)
+        out, err = capsys.readouterr()
+        assert 'Your passphrase (between double-quotes): "test_passphrase"' in err
+        assert "UTF-8 encoding in hex" in err
+
+        # Case 3: Non-ASCII passphrase
+        passphrase = "tëst_päṩṩ"
+        hex_value = passphrase.encode("utf-8").hex()
+        Passphrase.verification(passphrase)
+        out, err = capsys.readouterr()
+        assert hex_value in err
```

Comment:
We removed the special case for non-ascii, so this test is a bit off. Guess it should test with a ascii passphrase.